core/optaplanner-core-impl/src/main/java/org/optaplanner/core/config/AbstractConfig.java

```diff
@@ -51,4 +51,11 @@ public String toString() {
         return getClass().getSimpleName() + "()";
     }
 
+    /**
+     * Used as a return value of the with() methods on Config classes to ensure the correct type
+     * in the hierarchical builder pattern.
+     */
+    protected Config_ self() {
+        return (Config_) this;
+    }
```

Comment:
@rsynek Can you show me one place where returning `this` is not enough and using this method is necessary?

Reply:
Sure, return this method to its previous form:
https://github.com/kiegroup/optaplanner/pull/2675/files#diff-f87e82fb1dffc04b78e9c6843c192d5dd671fcc42b8e66c62216837470caa6eeR220

and have a look at the newly added test:
[core/optaplanner-core-impl/src/test/java/org/optaplanner/core/config/heuristic/selector/move/generic/PillarSwapMoveSelectorConfigTest.java](https://github.com/kiegroup/optaplanner/pull/2675/files#diff-63b7d8a5618e9a5b7ff9b7c8cfe2d33222cbb3dca90cb33e4b0d02b1be169224R11)

In the previous implementation, you can chain the `with` methods only from the child classes to parent classes, not the other way around, as the abstract parent class does not know the `with` methods of the child class.

Reply:
@yurloc please let me know if you still have concerns; I would like to merge the PR tomorrow.

Reply:
I understand the problem and I see why `MoveSelectorConfig`'s "generic" `with` methods need to return `Config_` and not `MoveSelectorConfig`.

I believed this could be achieved without using the `self()` method simply by returning `this` and I was asking for a counter example (proving my previously stated belief). The example you shared just illustrates the problem that I already understand, which confuses me a bit.

Although there might have been a bit of misunderstanding, you have removed the `self()` method, if I see correctly. That confirms what I thought was possible. I don't have any other concerns :slightly_smiling_face:

Reply:
I removed the `self()` method based on your and Lukas's feedback. To sum it up, `this` has to be cast to `Config_`; either in every return method or in a single place, which was the `self()` method.